### PR TITLE
Issue 452 add interactive connectivity map

### PR DESCRIPTION
**Pull request by haim-kermany**
main code for creating interactive map
still more work will be done later:
1. handling errors.
2. add relevant cli (command line, sceme)
3. updating document
4. explainability
5. informative legend
6. tooltip
7. and more...

### CODE REVIEW

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+            if t_class == self.CONNECTIVITY_CT:
+                tag['text-decoration'] = 'underline'
+                tag['font-weight'] = 'bold'
+
+        def _save_tag_file(self, tag_soup, tag_info):
+            """
+            save the soup to an svg file, the name and location is according to the class and id
+            (all svg file, except the main file, are in sub directory)
+            param: tag_soup: the soup to save: soup
+            param: tag_info: the information of the tag the we currently save its soup
```
Comment (zivnevo) [inline nca/FWRules/InteractiveConnectivityGraph.py:241]:
```suggestion
            param: tag_info: the information of the tag for which we currently save its soup
```
???

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+             param: t_class: the class of the tag that creates the svg file for: str
+            """
+            if (t_class == self.BACKGROUND_CT and related_tag_info.t_class == self.BACKGROUND_CT) or \
+               (t_class != self.BACKGROUND_CT and related_tag_info.t_class != self.BACKGROUND_CT):
+                relative_path = '.'
+            elif t_class == self.BACKGROUND_CT and related_tag_info.t_class != self.BACKGROUND_CT:
+                relative_path = 'elements'
+            else:
+                relative_path = '..'
+            related_tag['xlink:href'] = posixpath.join(relative_path, related_tag_info.t_id + '.svg')
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:216]: posixpath is '/' while os.path is '\' or '/'
the link in svg file wants always '/' on all platforms

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+            """
+            if tag_info.t_class == self.BACKGROUND_CT:
+                tag_file_name = os.path.join(self.output_directory, tag_info.t_id + '.svg')
+            else:
+                tag_file_name = os.path.join(self.output_directory, 'elements', tag_info.t_id + '.svg')
+            try:
+                with open(tag_file_name, 'wb') as tag_cvg_file:
+                    tag_cvg_file.write(tag_soup.prettify(encoding='utf-8'))
+            except Exception as e:
+                print(f'Failed to open file: {self.tag_cvg_file}\n{e} for writing', file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:251]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+            (all svg file, except the main file, are in sub directory)
+            param: tag_soup: the soup to save: soup
+            param: tag_info: the information of the tag the we currently save its soup
+            """
+            if tag_info.t_class == self.BACKGROUND_CT:
+                tag_file_name = os.path.join(self.output_directory, tag_info.t_id + '.svg')
+            else:
+                tag_file_name = os.path.join(self.output_directory, 'elements', tag_info.t_id + '.svg')
+            try:
+                with open(tag_file_name, 'wb') as tag_cvg_file:
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:248]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+            self.output_directory = output_directory
+            self.soup = None
+
+        def read_input_file(self):
+            """
+            read the file, and save is in soup object
+            """
+            try:
+                with open(self.input_svg_file) as cvg_file:
+                    self.soup = BeautifulSoup(cvg_file.read(), 'xml')
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:120]: done

File: nca/FWRules/InteractiveConnectivityGraph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply nca/FWRules/InteractiveConnectivityGraph.py:216]: ok, got it.

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+        @dataclass
+        class Graph:
+            t_id: str = ''
+            name: str = ''
+            namespaces: dict = field(default_factory=dict)
+            nodes: dict = field(default_factory=dict)
+            edges: dict = field(default_factory=dict)
+            cliques: list = field(default_factory=list)
+            bicliques: list = field(default_factory=list)
+            conn_legend: object = None
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:359]: great, done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+import networkx
+import os
+import sys
+import shutil
+import itertools
+import copy
+from dataclasses import dataclass, field
+from collections import defaultdict
+from bs4 import BeautifulSoup
+import posixpath
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:15]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+                if namespace:
+                    namespace.nodes.append(node)
+
+            for (src_name, dst_name), edge in self.graph.edges.items():
+                edge.src = self.graph.nodes[src_name]
+                edge.dst = self.graph.nodes[dst_name]
+
+            # creating cliques:
+            # find all cliques nodes and edges:
+            all_cliques_nodes = [node for node in self.graph.nodes.keys() if node.startswith('clique_')]
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:418]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+            all_cliques_edges = [edge for edge in itertools.product(all_cliques_nodes, all_cliques_nodes) if
+                                 edge in self.graph.edges.keys()]
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:420]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+                if tag[self.CLASS_TA] == self.EDGE_CT and tag.find('text'):
+                    tag[self.CONNECTIVITY_TA] = tag.find('text').string
+                elif tag[self.CLASS_TA] == self.NODE_CT and tag.title and 'clique' in tag.title.string:
+                    tag[self.CONNECTIVITY_TA] = tag.find_all('text')[1].string
+                elif tag[self.CLASS_TA] == self.CONNECTIVITY_CT:
+                    tag[self.CONNECTIVITY_TA] = tag[self.TITLE_TA]
+
+        def _get_soup_tag_info(self, tag):
+            """
+            read a the information from a soup tag
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:177]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+            """
+            t_id = tag[self.ID_TA]
+            t_class = tag[self.CLASS_TA]
+            t_title = tag[self.TITLE_TA]
+            t_conn = tag.get(self.CONNECTIVITY_TA) if tag.get(self.CONNECTIVITY_TA) else ''
+            return InteractiveConnectivityGraph.ElementInfo(t_id, t_class, str(t_title), t_conn)
+
+        def get_elements_info(self):
+            """
+            read a the information from all soup tags
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:189]: done

File: nca/FWRules/InteractiveConnectivityGraph.py (hunk 1 above)
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:241]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+            return elements_info
+
+        def _set_related_tag_link(self, related_tag, related_tag_info, t_class):
+            """
+            Set the link in the soup tag.
+            Not all the svg files sits in the same directory,
+            (all svg file, except the main file, are in sub directory)
+            therefore, relative path depends on the class of:
+             1. the element that we creates the svg file for
+             2. the soup tag the we currently update the link
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:203]: (meant to be 'that')
done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+                tag.polygon['stroke-width'] = '5'
+            if t_class == self.NAMESPACE_CT:
+                tag.polygon['stroke-width'] = '5'
+                tag['font-weight'] = 'bold'
+            if t_class == self.EDGE_CT:
+                tag.path['stroke-width'] = '3'
+                tag['font-weight'] = 'bold'
+            if t_class == self.CONNECTIVITY_CT:
+                tag['text-decoration'] = 'underline'
+                tag['font-weight'] = 'bold'
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:234]: done

**Comment by haim-kermany** [comment]
> This is so cool !!! Some comments to start with:
> 
>     * Need to add `beautifulsoup4==4.12.0` to `requirements.txt`
> 
>     * Need to allow `svg` as a legal output format and to make sure graphviz runs when SVG format is selected
> 
>     * Need to decide how user specifies the output directory for all graphs

updated requirements.txt
all other issues at another PR

File: nca/FWRules/InteractiveConnectivityGraph.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 14]
```diff
@@ -0,0 +1,547 @@
...
+                for edge in node.edges:
+                    elements_relations[node.t_id].relations |= elements_relations[edge.t_id].relations
+
+            # add edge and nodes to all clique elements (conns already inside)
+            # highlights all clique core elements with each other
+            for clique in self.graph.cliques:
+                for el in clique.nodes + clique.edges:
+                    for e in clique.edges:
+                        elements_relations[el.t_id].relations |= elements_relations[e.t_id].relations
+                        elements_relations[clique.conn.t_id].relations |= elements_relations[e.t_id].relations
```
Comment (zivnevo) [inline nca/FWRules/InteractiveConnectivityGraph.py:505]: Seems like we are adding to `elements_relations[clique.conn.t_id].relations` the same sets of edge relations over and over again. Can this be improved?

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+        def _get_soup_tag_info(self, tag):
+            """
+            read the information from a soup tag
+            param: tag: the tag from the soup object
+            return ElementInfo
+            """
+            t_id = tag[self.ID_TA]
+            t_class = tag[self.CLASS_TA]
+            t_title = tag[self.TITLE_TA]
+            t_conn = tag.get(self.CONNECTIVITY_TA) if tag.get(self.CONNECTIVITY_TA) else ''
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:185]: tag here is not a dict, but a soup element, and soup.get() does not support default value

File: tests/fw_rules_tests/policies/expected_output/poc1-scheme_output.dot
[the remark this answers is not in the document]
Comment (haim-kermany) [reply tests/fw_rules_tests/policies/expected_output/poc1-scheme_output.dot:1]: we should, there is an error there.
taking a look

[the remark this answers is not in the document]
Comment (haim-kermany) [reply tests/fw_rules_tests/policies/expected_output/poc1-scheme_output.dot:1]: not an error, handled in the next PR

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 16]
```diff
@@ -0,0 +1,547 @@
...
+                if os.path.isdir(self.output_directory):
+                    shutil.rmtree(self.output_directory)
+                os.mkdir(self.output_directory)
+                os.mkdir(os.path.join(self.output_directory, 'elements'))
+            except Exception as e:
+                print(f'Failed to create directory: {self.output_directory}\n{e} for writing', file=sys.stderr)
+                return
+            for tag in self.soup.svg.find_all('a'):
+                tag_info = self._get_soup_tag_info(tag)
+                tag_soup = copy.copy(self.soup)
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:280]: for each tag, we edit tag_soup - removing elements, highlight, to create a separate file.
we do not want to edit self.soup

File: tests/fw_rules_tests/policies/expected_output/poc1-scheme_output.dot
[the remark this answers is not in the document]
Comment (zivnevo) [reply tests/fw_rules_tests/policies/expected_output/poc1-scheme_output.dot:1]: ok

File: nca/FWRules/InteractiveConnectivityGraph.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply nca/FWRules/InteractiveConnectivityGraph.py:280]: ok, got it.

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 17]
```diff
@@ -0,0 +1,547 @@
...
+
+             param: related_tag: the tag from the soup object that we want to update its link
+             param: related_tag_info :the information of this tag: ElementInfo
+             param: t_class: the class of the tag that creates the svg file for: str
+            """
+            if (t_class == self.BACKGROUND_CT and related_tag_info.t_class == self.BACKGROUND_CT) or \
+               (t_class != self.BACKGROUND_CT and related_tag_info.t_class != self.BACKGROUND_CT):
+                relative_path = '.'
+            elif t_class == self.BACKGROUND_CT and related_tag_info.t_class != self.BACKGROUND_CT:
+                relative_path = 'elements'
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:214]: done

File: nca/FWRules/InteractiveConnectivityGraph.py [hunk 18]
```diff
@@ -0,0 +1,547 @@
...
+            # we find connected set of cliques nodes, each set is a clique
+            clqs_graph = networkx.Graph()
+            clqs_graph.add_nodes_from(all_cliques_nodes)
+            clqs_graph.add_edges_from(all_cliques_edges)
+            clique_sets = networkx.connected_components(clqs_graph)
+
+            # for each set build its clique, add its nodes and edges:
+            for clique_set in clique_sets:
+                cliqut_conn = self.graph.nodes[list(clique_set)[0]].conn
+                clique = self.Clique(cliqut_conn)
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:432]: done

File: nca/FWRules/InteractiveConnectivityGraph.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:185]: correct, sorry, done

File: nca/FWRules/InteractiveConnectivityGraph.py (hunk 14 above)
Comment (haim-kermany) [reply nca/FWRules/InteractiveConnectivityGraph.py:505]: done

**Comment by haim-kermany** [comment]
> A few more comments. Also, can you say something about the performance of this code? Seems like some parts of the code have O(|E|^3) complexity, which might be painful for large graphs.

I didnt pay any focus on complexity,  all the graphs that we have on the tests runs in ~sec.
I can handle it after I will run on otc graph

**Pull request merged**